src/dmd/typesem.d

```diff
@@ -3080,9 +3081,20 @@ private extern(C++) final class DotExpVisitor : Visitor
                 /* Rewrite e.ident as:
                  *  e.aliasthis.ident
                  */
-                e = resolveAliasThis(sc, e);
-                auto die = new DotIdExp(e.loc, e, ident);
-                return returnExp(die.semanticY(sc, flag & 1));
+                auto alias_e = resolveAliasThis(sc, e, gagError);
+
+                if (!alias_e)
+                    return returnExp(null);
+
+                auto die = new DotIdExp(e.loc, alias_e, ident);
+                auto exp = die.semanticY(sc, gagError);
+
+                if (exp && gagError)
+                    // now that we know that the alias this leads somewhere useful,
+                    // go back and print deprecations/warnings that we skipped earlier due to the gag
+                    resolveAliasThis(sc, e, false);
```

Comment:
Makes sense, but pretty awful that you have to call this twice.

Reply:
I agree. I don't know a better way though.